### ISSUE TITLE
How to make big image beyond the max size of CanvasBitmap.

### ISSUE THREAD

**Issue by omanuke** [Microsoft/Win2D#153]
I wanna make big image beyond the max size of CanvasBitmap ,but I don't know how to.
I have to compose large bitmap from the bitmap of each part drawn with CanvasBitmap?

**Comment by shawnhar** [comment]
To display very large scrollable images that are drawn in pieces, you can use a CanvasVirtualControl.   There is a demo of that in Example Gallery.

Win2D does not currently provide any way to load source bitmaps larger than the GPU hardware max texture size.  If you need very large source bitmaps, you will have to manually split them into pieces and draw each part separately.

An alternative is to use software rendering (http://microsoft.github.io/Win2D/html/P_Microsoft_Graphics_Canvas_UI_Xaml_CanvasControl_ForceSoftwareRenderer.htm) instead of the hardware GPU.   That's not available on Windows Phone versions < 10, and obviously won't perform as well as using the graphics hardware, but software rendering has vastly larger max texture size which can sometimes be useful.

**Comment by omanuke** [comment]
Sorry, my question is obscure.
I wanna create large size bitmap or other format. Not displaying.
I'm creating a kind of note app and use Win2D to display content. The size of content is almost no limit.
To display content, Win2D works very well. But to export content as large size image ,I don't know how to.

**Comment by shawnhar** [comment]
The only functionality in Win2D for saving out bitmaps is the CanvasBitmap.SaveAsync method.  Because this method saves out a bitmap, it is not possible for it to save images larger than the max bitmap size.  If you need to do that, you have two options:

1.  Switch from hardware GPU to a software graphics device, which will increase your max bitmap size  (available on Windows and Phone10, but not earlier Phone versions which did not include software rendering pipeline)

2.  Or you could use some other API than Win2D to save out the bitmap.  For instance you could interop to native code and call WIC APIs, which allow building up arbitrary large images in pieces.

**Comment by omanuke** [comment]
For me ,using software graphics device seems better.
Thx.

**Comment by omanuke** [comment]
I could create png with software rendering.
The size of resulted png is about 5500 pixel width and 3500 pixel height with 96 dpi as argument to new CanvasRenderTarget. But when I passed 192 dpi as construct parameter, SaveAsync method rasied OutOfMemory exception. 
My machine specs are Core i 7 CPU,16G memory, and GTX970 GPU.
What kind of memory was not enough?

**Comment by shawnhar** [comment]
On my dev PC I was able to successfully save out a 11,000 x 7,000 image (the same as you are doing here - changing DPI is basically the same thing as changing resolution) to both .png and .jpg formats.  But going much larger than this I also hit out of memory.

This is to be expected when working with such a large image.  At this resolution you are using several hundred megabytes just for a single copy of the pixel data, and multiple copies are needed at various points while manipulating the image (rendering to it, or encoding it to a file).

I don't really think it is feasible to treat such extremely large images as a single monolithic bitmap using a simple API like Win2D.  Programs that need to manipulate images of this scale generally need to split them up, processing smaller pieces separately and taking great care to unload one section from memory before moving on to the next.  That is out of scope for what Win2D can help you with, but the lower level WIC (Windows Imaging Component) APIs include advanced functionality for encoding images piece-by-piece.

**Comment by rizamarhaban** [comment]
Okay, help!

Sorry for asking directly in here as I don't think people in StackOverflow have the better answer and I have deadline projects that I need to finish where I promoted to use Win2D in UWP and all works fine, until...

Short story: We have huge size 2D drawing from AutoCAD and we would like to animate the drawing inside UWP app. What could go wrong, I use Win2D and it works perfectly. And since the drawing has lots of entities, we manage to create a converter from AutoCAD to SVG smoothly. No Issue with this. However, once I use this,

```c#
CanvasSvgDocument.LoadAsync(...)
```

Let me show my code:

```c#
private async Task<CanvasBitmap> CreateBitmapFromSvg(CanvasAnimatedControl sender, SVGDocument svg)
{
    CanvasBitmap bitmap = null;

    var size = new Windows.Foundation.Size(svg.Size.Width, svg.Size.Height); // fake size, not using viewport
    var svgImage = await CanvasSvgDocument.LoadAsync(sender, svg.GetSVGStream().AsRandomAccessStream());

    using (var stream = new InMemoryRandomAccessStream())
    {
        var rt= new CanvasRenderTarget(sender.Device, svg.Size.Width, svg.Size.Height, 96);

        using (CanvasDrawingSession ds= rt.CreateDrawingSession())
        {
            ds.DrawSvg(svgImage, size);
        }

        await rt.SaveAsync(stream, CanvasBitmapFileFormat.Png);  // <---- save first
        await rt.SaveAsync($"{_cache.Path}\\{svg.Name}.png", CanvasBitmapFileFormat.Png, 1f);  // <---- this saved file was cropped

        defaultRenderTarget.Dispose();

        bitmap = await CanvasBitmap.LoadAsync(sender, stream);  // <---- this also cropped
    }
    return bitmap;
}
```

will work if the size is less than 8192 pixels. I get it, due to hardware limitations. My Laptop GPU memory sucks. Then, I also want to try using `CanvasVirtualBitmap` to save the image but how do I draw the image using `CanvasDrawingSession` before I can do `CanvasVirtualBitmap.LoadAsync(...)` and `CanvasImage.SaveAsync(...)`. Point me to an example would be awesome. Cannot find one from the existing samples.

Question, How do I solve this? Is there a better way since I can load and stitch the still image once generated and use for the background on the animation. I have no issue with the animation as you can view the video on my twitter: https://twitter.com/rizamarhaban/status/1155437434319933440

Question, is there a way to crop directly from CanvasSvgDocument before we can draw it? I was hoping there is a method like this:

```c#
        using (CanvasDrawingSession ds= rt.CreateDrawingSession())
        {
             // wishing have this. Clipping directly from CanvasSVGDocument source.
            ds.DrawSvg(svgImage, size, new Rect(... clip bound source from svgImage...));
        }
```

**Comment by shawnhar** [comment]
You cannot use a CanvasRenderTarget if you want to go larger than the maximum possible rendertarget size.  Instead, draw the SVG into a CanvasCommandList, then use CanvasImage.SaveAsync.
